app/models/cart.rb

```diff
@@ -26,20 +26,41 @@ def subtotal(item)
     item.price * @contents[item.id.to_s]
   end
 
+  def discount_subtotal(item)
+    item.discount_price * @contents[item.id.to_s]
+  end
+
   def total
     @contents.sum do |item_id,quantity|
       Item.find(item_id).price * quantity
     end
   end
 
   def discount_total
-    items.sum do |item, quantity|
-      discount = Discount.find_by(merchant_id: item.merchant_id)
-      if item.merchant_id == discount.merchant_id && quantity == discount.item_quantity
-        (item.price - (item.price / discount.percentage_discount)) * quantity
-      else
-        item.price * quantity
+    if discount_applies
+      @contents.sum do |item_id, quantity|
+        if has_discount(Item.find_by(id: item_id))
+          Item.find(item_id).discount_price * quantity
+        else
+          Item.find(item_id).price * quantity
+        end
       end
     end
   end
+
+  def has_discount(item)
+    !Discount.find_by(merchant_id: item.merchant_id).nil? && Discount.find_by(merchant_id: item.merchant_id).item_quantity <= items[item]
+  end
+
+  def discount(item)
+    if has_discount(item)
+      Discount.find_by(merchant_id: item.merchant_id)
+    end
+  end
+
+  def discount_applies
+    @contents.any? do |item_id|
```

Comment:
Did my best to use AR. The research I did seemed to suggest that this was a valid way of doing so.